vsphere/stackstate_checks/vsphere/vsphere.py

```diff
@@ -509,38 +503,57 @@ def _get_server_instance(self, instance):
+            )
+        except Exception as e:
+            err_msg = "Connection to {} failed: {}".format(instance.get('host'), e)
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=service_check_tags, message=err_msg)
+            raise Exception(err_msg)
+
+        # Check that we have sufficient permission for the calls we need to make
+        try:
+            server_instance.CurrentTime()
+        except Exception as e:
+            err_msg = (
+                "A connection to {} can be established, but performing operations on the server fails: {}"
+            ).format(instance.get('host'), e)
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL, tags=service_check_tags, message=err_msg)
+            raise Exception(err_msg)
+
+        return server_instance
+
+    def _get_server_instance(self, instance):
+        i_key = self._instance_key(instance)
+
+        service_check_tags = [
+            'vcenter_server:{0}'.format(instance.get('name')),
+            'vcenter_host:{0}'.format(instance.get('host')),
+        ]
 
         if i_key not in self.server_instances:
-            try:
-                # Object returned by SmartConnect is a ServerInstance
-                #   https://www.vmware.com/support/developer/vc-sdk/visdk2xpubs/ReferenceGuide/vim.ServiceInstance.html
-                server_instance = connect.SmartConnect(
-                    host=instance.get('host'),
-                    user=instance.get('username'),
-                    pwd=instance.get('password'),
-                    connectionPoolTimeout=-1,
-                    sslContext=context if not ssl_verify or ssl_capath else None
-                )
-            except Exception as e:
-                err_msg = "Connection to %s failed: %s" % (instance.get('host'), e)
-                self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
-                                   tags=service_check_tags, message=err_msg)
-                raise Exception(err_msg)
-            self.server_instances[i_key] = server_instance
+            self.server_instances[i_key] = self._smart_connect(instance, service_check_tags)
 
         # Test if the connection is working
         try:
-            self.server_instances[i_key].RetrieveContent()
+            self.server_instances[i_key].CurrentTime()
             self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK, tags=service_check_tags)
-        except Exception as e:
-            err_msg = "Connection to %s died unexpectedly: %s" % (instance.get('host'), e)
-            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
-                               tags=service_check_tags, message=err_msg)
-            raise Exception(err_msg)
+        except Exception:
+            # Try to reconnect, if the connection is definitely broken.
+            self.server_instances[i_key] = self._smart_connect(instance, service_check_tags)
```

Comment:
If this reconnect also fails we should drop `i_key` from the `self.server_instances` map, otherwise we keep the "broken" connection in the state

Reply:
@zandrewitte we are overriding it anyway, so it won't matter I think.

Reply:
If `self._smart_connect(instance, service_check_tags)` fails and throws an exception on either the `connect.SmartConnect` or `server_instance.CurrentTime()` we won't override it and that connection will remain in `self.server_instances`. So next time the check runs it will fail again on  `self.server_instances[i_key].CurrentTime()`.

Does that make sense? It won't blow up, but it will fail once unnecessarily

Reply:
ok. done.